riscv: telink:tl3218x: save ram by change config.

haiwentelink opened the pull request:
- add special config for tl3218x.

haiwentelink commented:
need to adjust linkscript for zephyr first .

liang4804 commented:
The previous changes of mine are basically covered.

s07641069 commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -255,6 +285,7 @@ config NVS_LOOKUP_CACHE
 
 config NVS_LOOKUP_CACHE_SIZE
     default 2048 if !PM
+    default 400 if SOC_RISCV_TELINK_TL321X
```
check .config after build which value will be selected
because "if !PM" is first one and also will be valid condition
In case if 2048 will be selected in .config (build folder) just move "default 400 if SOC_RISCV_TELINK_TL321X" into first place

haiwentelink replied:
yes , i think so , will change , move to the first place.

haiwentelink replied:
```
config NVS_LOOKUP_CACHE_SIZE
    default 400  if SOC_RISCV_TELINK_TL321X
    default 2048 if !PM
```

s07641069 commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -62,24 +62,28 @@ config POSIX_MAX_FDS
 # Application stack size
 config MAIN_STACK_SIZE
     default 3240 if PM
+    default 4096 if SOC_RISCV_TELINK_TL321X
```
make no sense to add here, because value is same as "default 4096"

haiwentelink replied:
it's a mistake , I will change it , actually the value is 3240 .

haiwentelink replied:
```
config MAIN_STACK_SIZE
    default 3240 if PM || SOC_RISCV_TELINK_TL321X
    default 4096
```

s07641069 commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -62,24 +62,28 @@ config POSIX_MAX_FDS
 
 config IDLE_STACK_SIZE
     default 512 if PM
+    default 512 if SOC_RISCV_TELINK_TL321X
```
In case if default value for "SOC_RISCV_TELINK_TL321X" is same as for "PM" we can use next condition:
```
config IDLE_STACK_SIZE
    default 512 if PM || SOC_RISCV_TELINK_TL321X
```

In case if default value for "SOC_RISCV_TELINK_TL321X" is different then for "PM" we need to understand that in feature SOC_RISCV_TELINK_TL321X will have PM as well and you will have conflict.
to avoid this put default value condition which has lower value alway into first place.

Same for all rest changes in this PR...

haiwentelink replied:
OK, will change .

haiwentelink replied:
```
config IDLE_STACK_SIZE
    default 512 if PM || SOC_RISCV_TELINK_TL321X
```
also proc others .

s07641069 commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -129,17 +137,39 @@ config BT_MAX_CONN
+    bool
+    default n
+
+config IPC_SERVICE
```
Are you sure that "config IPC_SERVICE" is enabled by default for others SoCs except w91 (check in build .config) ?

haiwentelink replied:
@liang4804  need to check by liang . temporary , i will delete it .

s07641069 commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -129,17 +137,39 @@ config BT_MAX_CONN
 config BT_RX_STACK_SIZE
     default 810 if BT_B9X
     default 810 if BT_TLX
+    default 1010 if SOC_RISCV_TELINK_TL321X
```
Here you can see that I already added condition for new TLX SoC
`default 810 if BT_TLX`
Please recheck and remove your to avoid conflict

haiwentelink replied:
OK, change the default  "default 1010 if BT_TLX"

haiwentelink replied:
```
config BT_RX_STACK_SIZE
    default 810 if BT_B9X
    default 1010 if BT_TLX
    default 2048 if BT_W91
```

s07641069 commented on src/platform/telink/CHIPPlatformConfig.h:
```diff
@@ -72,6 +72,41 @@
+#define CHIP_CONFIG_DEVICE_MAX_ACTIVE_DEVICES 2
+#endif // CHIP_CONFIG_DEVICE_MAX_ACTIVE_DEVICES
+
+#elif defined CONFIG_SOC_RISCV_TELINK_TL321X
```
same here what we will have in case if PM will be enabled together with CONFIG_SOC_RISCV_TELINK_TL321X
As I see only next two configs (CHIP_CONFIG_MAX_GROUP_DATA_PEERS, CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC) has different values then under #ifdef CONFIG_PM

I suggest to do next, define firstly lower values which are different and after define same values for both conditions (or provide you valiant):

#if defined CONFIG_SOC_RISCV_TELINK_TL321X

#ifndef CHIP_CONFIG_MAX_GROUP_DATA_PEERS
#define CHIP_CONFIG_MAX_GROUP_DATA_PEERS 5
#endif // CHIP_CONFIG_MAX_GROUP_DATA_PEERS

#ifndef CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC
#define CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC 4
#endif // CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC

#endif // CONFIG_SOC_RISCV_TELINK_TL321X

#if defined CONFIG_PM || defined CONFIG_SOC_RISCV_TELINK_TL321X

#ifndef CHIP_CONFIG_MAX_GROUP_DATA_PEERS
#define CHIP_CONFIG_MAX_GROUP_DATA_PEERS 7
#endif // CHIP_CONFIG_MAX_GROUP_DATA_PEERS
#ifndef CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC
#define CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC 2
#endif // CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC

#ifndef CHIP_IM_MAX_NUM_COMMAND_HANDLER
#define CHIP_IM_MAX_NUM_COMMAND_HANDLER 1
#endif // CHIP_IM_MAX_NUM_COMMAND_HANDLER

#ifndef CHIP_IM_MAX_NUM_WRITE_HANDLER
#define CHIP_IM_MAX_NUM_WRITE_HANDLER 1
#endif // CHIP_IM_MAX_NUM_WRITE_HANDLER

#ifndef CHIP_IM_MAX_NUM_WRITE_CLIENT
#define CHIP_IM_MAX_NUM_WRITE_CLIENT 1
#endif // CHIP_IM_MAX_NUM_WRITE_CLIENT

#ifndef CHIP_IM_MAX_REPORTS_IN_FLIGHT
#define CHIP_IM_MAX_REPORTS_IN_FLIGHT 1
#endif // CHIP_IM_MAX_REPORTS_IN_FLIGHT

#ifndef CHIP_CONFIG_LAMBDA_EVENT_SIZE
#define CHIP_CONFIG_LAMBDA_EVENT_SIZE (16)
#endif // CHIP_CONFIG_LAMBDA_EVENT_SIZE

#ifndef CHIP_CONFIG_DEVICE_MAX_ACTIVE_DEVICES
#define CHIP_CONFIG_DEVICE_MAX_ACTIVE_DEVICES 2
#endif // CHIP_CONFIG_DEVICE_MAX_ACTIVE_DEVICES

#endif // CONFIG_PM || CONFIG_SOC_RISCV_TELINK_TL321X

haiwentelink replied:
OK, by this code will be more clean , the initial setting which we want to set the same value with PM'settings . extra part can by define .

haiwentelink replied:
```
#ifdef CONFIG_SOC_RISCV_TELINK_TL321X

#ifndef CHIP_CONFIG_MAX_GROUP_DATA_PEERS
#define CHIP_CONFIG_MAX_GROUP_DATA_PEERS 5
#endif // CHIP_CONFIG_MAX_GROUP_DATA_PEERS

#ifndef CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC
#define CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC 4
#endif // CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC

#endif // CONFIG_SOC_RISCV_TELINK_TL321X
```

haiwentelink replied:
```
#if defined CONFIG_PM || defined CONFIG_SOC_RISCV_TELINK_TL321X

#ifndef CHIP_CONFIG_MAX_GROUP_DATA_PEERS
#define CHIP_CONFIG_MAX_GROUP_DATA_PEERS 7
#endif // CHIP_CONFIG_MAX_GROUP_DATA_PEERS

#ifndef CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC
#define CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC 2
#endif // CHIP_CONFIG_EXAMPLE_ACCESS_CONTROL_MAX_ENTRIES_PER_FABRIC

#ifndef CHIP_IM_MAX_NUM_COMMAND_HANDLER
#define CHIP_IM_MAX_NUM_COMMAND_HANDLER 1
#endif // CHIP_IM_MAX_NUM_COMMAND_HANDLER
#ifndef CHIP_IM_MAX_NUM_WRITE_HANDLER
#define CHIP_IM_MAX_NUM_WRITE_HANDLER 1
#endif // CHIP_IM_MAX_NUM_WRITE_HANDLER
#ifndef CHIP_IM_MAX_NUM_WRITE_CLIENT
#define CHIP_IM_MAX_NUM_WRITE_CLIENT 1
#endif // CHIP_IM_MAX_NUM_WRITE_CLIENT
#ifndef CHIP_IM_MAX_REPORTS_IN_FLIGHT
#define CHIP_IM_MAX_REPORTS_IN_FLIGHT 1
#endif // CHIP_IM_MAX_REPORTS_IN_FLIGHT
#ifndef CHIP_CONFIG_LAMBDA_EVENT_SIZE
#define CHIP_CONFIG_LAMBDA_EVENT_SIZE (16)
#endif // CHIP_CONFIG_LAMBDA_EVENT_SIZE
#ifndef CHIP_CONFIG_DEVICE_MAX_ACTIVE_DEVICES
#define CHIP_CONFIG_DEVICE_MAX_ACTIVE_DEVICES 2
#endif // CHIP_CONFIG_DEVICE_MAX_ACTIVE_DEVICES

#endif // CONFIG_PM || CONFIG_SOC_RISCV_TELINK_TL321X
```

s07641069 commented on src/platform/telink/SystemPlatformConfig.h:
```diff
@@ -44,14 +44,26 @@ struct ChipDeviceEvent;
 #ifdef CONFIG_PM
 #define CHIP_SYSTEM_CONFIG_PACKETBUFFER_POOL_SIZE 0
 #else
+    #ifdef CONFIG_SOC_RISCV_TELINK_TL321X
```
same here, values are equal
you can use next condition:
```
#if defined CONFIG_PM || defined CONFIG_SOC_RISCV_TELINK_TL321X
#define CHIP_SYSTEM_CONFIG_PACKETBUFFER_POOL_SIZE 0
#else
#define CHIP_SYSTEM_CONFIG_PACKETBUFFER_POOL_SIZE 8
#endif
```

haiwentelink replied:
aggree

s07641069 commented on src/platform/telink/tl3218x_2m_flash.overlay:
```diff
@@ -18,7 +18,7 @@
 		factory_partition: partition@102000 {
 			label = "factory-data";
-			reg = <0x102000 0x1000>;
+			reg = <0x102000 0x800>;
```
After decreasing of factory_partition you will have 0x800 of unused space
As you can see in all others flash overlays we always used 100% of flash memory
Please redistribute unused space into other slots where it more needed, for example add 0x400 for each slot0/1

haiwentelink replied:
I don't think so , for the slot0 when we want to operate , it will need to erase which the size is 4k , it's better to create another RFU section . I will change .

s07641069 commented:
@haiwentelink 
As you can see under branch [telink_tl321x](https://github.com/s07641069/connectedhomeip/tree/telink_tl321x) I added GitHub CI builds for tl321x and tl321x_retention (you can disable retention build temporary in case of additional fails)
See .github/workflows/examples-telink.yaml
There is additional step which I uncommented:
```
- name: Update Zephyr to specific revision (for developers purpose)
  run: scripts/run_in_build_env.sh "python3 scripts/tools/telink/update_zephyr.py 322432f1d903337f9e5501835d7057a7a73168f6" 
```
Please update time to time Zephyr revision into latest HASH for tl321x, this will help to check build by CI with your latest Zephyr/HAL changes

s07641069 commented:
@haiwentelink 
Also check my changes under parent branch here diff https://github.com/project-chip/connectedhomeip/compare/master...s07641069:connectedhomeip:telink_tl321x?expand=1
If some defaults values are not ok, just add modifications in your PR as well as some prepared, but commented code by me in my branch

haiwentelink commented:
zephyr commit id have already include ,and the compile for deep-retention have already removed .let's wait the result of CI 
       Actually , the buteo will not used in pm project .

s07641069 commented:
@haiwentelink I removed retention
In case if there no retention and no PM you may clean/remove all retention related code/configs from Zephyr as well or it will be used for others SoCs in same SoC family in feature?

haiwentelink committed:
riscv: telink:tl3218x: clean code and fix CI.

- fix CI error.
- clean code and improve settings.
- delete retention setting for TL321X.

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>

haiwentelink commented:
@s07641069  I think this version can be merged which can commission suc by thread mode . After this ,we will have a version with ble .

s07641069 committed:
[Telink] Revert add retention board
